src/typing_manager.ts

```diff
@@ -135,57 +118,109 @@ export class TypingManager implements Disposable {
                 this.isEnteringInsertMode = true;
                 documentPromise.then(async () => {
                     await this.main.cursorManager.waitForCursorUpdate(editor);
-                    if (this.main.modeManager.isInsertMode) {
-                        this.disposeType();
-                        this.disposeReplacePrevChar();
+                    if (this.isInsertMode) {
+                        this.takeOverVSCodeInput = false;
                     }
                     if (this.pendingKeysAfterEnter) {
                         logger.debug(
                             `Replaying pending keys after entering insert mode: ${this.pendingKeysAfterEnter}`,
                         );
-                        await commands.executeCommand(this.main.modeManager.isInsertMode ? "default:type" : "type", {
+                        await commands.executeCommand(this.isInsertMode ? "default:type" : "type", {
                             text: this.pendingKeysAfterEnter,
                         });
                         this.pendingKeysAfterEnter = "";
                     }
                     this.isEnteringInsertMode = false;
                 });
             } else {
-                this.disposeType();
-                this.disposeReplacePrevChar();
+                this.takeOverVSCodeInput = false;
             }
-        } else if (!this.main.modeManager.isInsertMode) {
+        } else if (!this.isInsertMode) {
             this.isEnteringInsertMode = false;
             this.isExitingInsertMode = false;
-            this.registerType();
-            this.registerReplacePrevChar();
+            this.takeOverVSCodeInput = true;
         }
     };
 
-    private onVSCodeType = async (_editor: TextEditor, edit: TextEditorEdit, type: { text: string }): Promise<void> => {
-        if (this.isEnteringInsertMode) {
-            this.pendingKeysAfterEnter += type.text;
-        } else if (this.isExitingInsertMode) {
-            this.pendingKeysAfterExit += type.text;
-        } else if (this.isInComposition) {
-            this.composingText += type.text;
-        } else if (this.main.modeManager.isInsertMode && !this.main.modeManager.isRecordingInInsertMode) {
-            if ((await this.client.mode).blocking) {
-                this.client.input(normalizeInputString(type.text, !this.main.modeManager.isRecordingInInsertMode));
+    async compositeInput(key: string) {
+        if (!this.compositeMatchedFirstKey) {
+            if (this.compositeFirstKeys.includes(key)) {
+                this.compositeMatchedFirstKey = key;
+                this.compositePromise?.resolve();
```

Comment:
If we're already concerned about the possibility of `compositePromise` being non-null, should we call `clearTimer` if the composite timer is non-null as well?

Reply:
`await null` and `clearTimeout(null)` are all valid

Reply:
Ah cool - then it would be a good idea to do `clearTimeout(this.compositeTimer)` here :)

Reply:
Here, timer cannot be non-null

Reply:
I removed the promise now 😂